### PR TITLE
[Enhancement] `files()` type promotion

### LINKED ISSUE #38992: [schema auto-detection merge] when query file table function the data file has different decimal type of parquet format, error
### Steps to reproduce the behavior (Required)
different deciaml type 
d41db890-467d-446f-942e-9e54af8e404a
ctas failed
image

### Expected behavior (Required)
execute ctas successfully
### Real behavior (Required)
execute ctas failed
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
mysql> select current_version();
+--------------------+
| current_version()  |
+--------------------+
| branch-3.2-68aaa24 |
+--------------------+
1 row in set (0.01 sec)

### PR DESCRIPTION
**Pull request by rickif**
## Why I'm doing:
The `files()` would merge file schemas. The current rule of merging is too simple.
## What I'm doing:
This PR makes some improvements to the `files` type promotion, including the following changes:
1. promote decimal types to bigger decimal type
2. promote string types to bigger string type
3. promote float type and integer type to double type
4. promote different complex types to vachar. 
Fixes https://github.com/StarRocks/starrocks/issues/38992

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/test/formats/orc/orc_chunk_reader_test.cpp [hunk 1]
```diff
@@ -2241,7 +2241,7 @@ TEST_F(OrcChunkReaderTest, get_file_schema) {
               {"col_float", TypeDescriptor::from_logical_type(TYPE_FLOAT)},
               {"col_double", TypeDescriptor::from_logical_type(TYPE_DOUBLE)},
               {"col_string", TypeDescriptor::create_varchar_type(1048576)},
-              {"col_char", TypeDescriptor::create_char_type(10)},
+              {"col_char", TypeDescriptor::create_char_type(255)},
```
Comment (dirtysalt) [inline be/test/formats/orc/orc_chunk_reader_test.cpp:2244]: use MAX_CHAR_LENGTH better?

File: be/src/runtime/types.h [hunk 2]
```diff
@@ -339,6 +346,39 @@ struct TypeDescriptor {
...
+            auto len = type1.len > type2.len ? type1.len : type2.len;
+            return TypeDescriptor::create_varbinary_type(len);
+        }
+        // treat other conflicted types as varchar.
+        return TypeDescriptor::create_varchar_type(TypeDescriptor::MAX_VARCHAR_LENGTH);
+    }
+
 private:
     /// Used to create a possibly nested type from the flattened Thrift representation.
     ///
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply be/src/runtime/types.h:388]: Does this comment help you ?

[the remark this answers is not in the document]
Comment (rickif) [reply be/src/runtime/types.h:388]: @dirtysalt Yeah. I've changed the PR according to the comment.

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by rickif** [comment]
@mergifyio rebase

**Review by wyb — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**